File: web/src/types/employee-evidence.ts
import type { AnalysisFilters, ValidationFinding } from '@/types/analysis'

export type EvidenceKpi = 'productivity' | 'compliance' | 'quality'

interface EvidenceRecord {
  record_id: string
  employee_id: string
}
export interface WorkOutputRecord extends EvidenceRecord {
  assigned_date: string
  due_date: string
  completed_date: string | null
  completion_status: string
  actual_effort_hours: number | null
  verification_status: string
  evidence_link: string | null
}
export interface AttendanceRecord extends EvidenceRecord {
  occurred_on: string
  outcome: string
  record_status: string
  scheduled_start: string | null
  actual_start: string | null
  lunch_out: string | null
  lunch_in: string | null
  scheduled_end: string | null
  actual_end: string | null
  confidence_score: number
}
export interface SubmissionRecord extends EvidenceRecord {
  due_date: string
  submitted_date: string | null
  outcome: string
  completeness_ratio: number
  verification_status: string
}
export interface LeaveRecord extends EvidenceRecord {
  category: string
  start_date: string
  end_date: string
  outcome: string
  documentation_complete: boolean
}
export interface QualityRecord extends EvidenceRecord {
  related_output_id: string
  occurred_on: string
  accuracy_ratio: number
  first_pass_approved: boolean
  rework_hours: number
  verification_status: string
}
interface EvidenceRow<Type, Record> {
  record_type: Type
  record_id: string
  record: Record
  validation_findings: ValidationFinding[]
  excluded_from_scoring: boolean
  exclusion_reason: string | null
}
export type WorkOutputRow = EvidenceRow<'work_output', WorkOutputRecord>
export type AttendanceRow = EvidenceRow<'attendance', AttendanceRecord>
export type SubmissionRow = EvidenceRow<'submission', SubmissionRecord>
export type LeaveRow = EvidenceRow<'leave', LeaveRecord>
export type QualityRow = EvidenceRow<'quality', QualityRecord>
export type ComplianceRow = AttendanceRow | SubmissionRow | LeaveRow
export type EmployeeEvidenceRow = WorkOutputRow | ComplianceRow | QualityRow
export interface EvidenceTable<Row = EmployeeEvidenceRow> {
  total_count: number
  rows: Row[]
}
export interface EmployeeEvidenceTables {
  productivity: EvidenceTable<WorkOutputRow>
  compliance: EvidenceTable<ComplianceRow>
  quality: EvidenceTable<QualityRow>
}
export interface EmployeeEvidenceResponse extends EvidenceTable {
  employee: { employee_id: string, employee_name: string | null, team: string | null, role: string | null }
  kpi: EvidenceKpi
  applied_filters: AnalysisFilters
  latest_submission_at: string
  page: number
  page_size: number
}

File: app/schemas/employee_evidence.py
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from app.schemas.performance import (
    AttendanceComplianceEvidence,
    Employee,
    LeaveComplianceEvidence,
    QualityEvidence,
    SubmissionComplianceEvidence,
    ValidationFinding,
    WorkOutputEvidence,
)
from app.schemas.uploads import AnalysisFilters

type EvidenceKpi = Literal["productivity", "compliance", "quality"]


class EvidenceRow(BaseModel):
    record_id: str
    validation_findings: list[ValidationFinding]
    excluded_from_scoring: bool = False
    exclusion_reason: str | None = None


class WorkOutputRow(EvidenceRow):
    record_type: Literal["work_output"] = "work_output"
    record: WorkOutputEvidence


class AttendanceRow(EvidenceRow):
    record_type: Literal["attendance"] = "attendance"
    record: AttendanceComplianceEvidence


class SubmissionRow(EvidenceRow):
    record_type: Literal["submission"] = "submission"
    record: SubmissionComplianceEvidence


class LeaveRow(EvidenceRow):
    record_type: Literal["leave"] = "leave"
    record: LeaveComplianceEvidence


class QualityRow(EvidenceRow):
    record_type: Literal["quality"] = "quality"
    record: QualityEvidence


type ComplianceRow = Annotated[
    AttendanceRow | SubmissionRow | LeaveRow, Field(discriminator="record_type")
]
type EmployeeEvidenceRow = Annotated[
    WorkOutputRow | AttendanceRow | SubmissionRow | LeaveRow | QualityRow,
    Field(discriminator="record_type"),
]


class EvidenceTable[Row](BaseModel):
    total_count: int = Field(ge=0)
    rows: list[Row]


class EmployeeEvidenceTables(BaseModel):
    productivity: EvidenceTable[WorkOutputRow]
    compliance: EvidenceTable[ComplianceRow]
    quality: EvidenceTable[QualityRow]


class EmployeeEvidenceResponse(BaseModel):
    employee: Employee
    kpi: EvidenceKpi
    applied_filters: AnalysisFilters
    latest_submission_at: datetime
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=50)
    total_count: int = Field(ge=0)
    all_records_count: int = Field(ge=0)
    needs_review_count: int = Field(ge=0)
    review_only: bool = False
    rows: list[EmployeeEvidenceRow]
